File: BAEKJOON/17143 - 낚시왕/낚시왕.cs
using System;
using System.Collections.Generic;

class Shark {
    public int speed;
    public int direction;
    public int size;
}

class Program {
    static int[] dr = { -1, 1, 0, 0 }; // 상하좌우 이동 방향
    static int[] dc = { 0, 0, 1, -1 };
    
    static int R, C, M; // 격자 크기와 상어 수
    static Dictionary<(int, int), Shark> sharks = new Dictionary<(int, int), Shark>(); // 상어 정보 저장
    
    // 상어 이동 함수
    static void MoveSharks() {
        Dictionary<(int, int), Shark> newSharks = new Dictionary<(int, int), Shark>(); // 이동 후 상어 저장

        foreach (var entry in sharks) {
            int r = entry.Key.Item1;
            int c = entry.Key.Item2;
            Shark shark = entry.Value;

            int speed = shark.speed;
            int direction = shark.direction;
            int size = shark.size;

            int nr = r, nc = c;

            // 위, 아래 이동 처리
            if (direction == 1 || direction == 2) {
                speed %= (R - 1) * 2; // 주기성을 이용한 속도
                for (int i = 0; i < speed; ++i) {
                    if (nr == 0 && direction == 1) direction = 2;
                    else if (nr == R - 1 && direction == 2) direction = 1;
                    nr += dr[direction - 1];
                }
            }
            // 좌우 이동 처리
            else {
                speed %= (C - 1) * 2; // 주기성을 이용한 속도
                for (int i = 0; i < speed; ++i) {
                    if (nc == 0 && direction == 4) direction = 3;
                    else if (nc == C - 1 && direction == 3) direction = 4;
                    nc += dc[direction - 1];
                }
            }

            // 상어가 이동 후 겹치면 가장 큰 상어만 남기기
            if (newSharks.ContainsKey((nr, nc))) {
                if (newSharks[(nr, nc)].size < size) {
                    newSharks[(nr, nc)] = new Shark { speed = speed, direction = direction, size = size };
                }
            } else {
                newSharks[(nr, nc)] = new Shark { speed = speed, direction = direction, size = size };
            }
        }

        sharks = newSharks; // 상어 상태 업데이트
    }

    static void Main() {
        string[] input = Console.ReadLine().Split();
        R = int.Parse(input[0]);
        C = int.Parse(input[1]);
        M = int.Parse(input[2]);

        // 상어 정보 입력
        for (int i = 0; i < M; i++) {
            input = Console.ReadLine().Split();
            int r = int.Parse(input[0]) - 1;
            int c = int.Parse(input[1]) - 1;
            int s = int.Parse(input[2]);
            int d = int.Parse(input[3]);
            int z = int.Parse(input[4]);
            sharks[(r, c)] = new Shark { speed = s, direction = d, size = z };
        }

        int totalSize = 0;

        // 낚시왕이 1번 열부터 마지막 열까지 이동
        for (int kingPos = 0; kingPos < C; kingPos++) {
            // 해당 열에서 가장 가까운 상어 잡기
            for (int row = 0; row < R; row++) {
                if (sharks.ContainsKey((row, kingPos))) {
                    totalSize += sharks[(row, kingPos)].size; // 상어 크기 합산
                    sharks.Remove((row, kingPos)); // 상어 제거
                    break;
                }
            }

            // 상어 이동 처리
            MoveSharks();
        }

        // 결과 출력
        Console.WriteLine(totalSize);
    }
}

File: BAEKJOON/17143 - 낚시왕/낚시왕.py
from collections import defaultdict

# 격자의 크기와 상어의 수를 입력받는다.
R, C, M = map(int, input().split())

# 상어 정보를 저장할 딕셔너리. 키는 (r, c) 좌표, 값은 [속도 s, 방향 d, 크기 z].
sharks = {}
for _ in range(M):
    r, c, s, d, z = map(int, input().split())
    sharks[(r-1, c-1)] = [s, d, z]  # 입력 좌표는 1부터 시작하므로 0-based 좌표로 변경하여 저장

# 상하좌우 이동을 위한 방향 배열 (1: 위, 2: 아래, 3: 오른쪽, 4: 왼쪽)
dr = [-1, 1, 0, 0]  # 행 이동 (위로 -1, 아래로 +1)
dc = [0, 0, 1, -1]  # 열 이동 (오른쪽 +1, 왼쪽 -1)

# 상어가 이동하는 함수
def move_sharks():
    # 새로 이동된 상어의 정보를 저장할 딕셔너리
    new_sharks = defaultdict(list)

    # 모든 상어에 대해 이동을 처리
    for (r, c), (s, d, z) in sharks.items():
        nr, nc, nd = r, c, d  # 현재 상어의 좌표와 방향을 가져옴

        # 상하 이동(1: 위, 2: 아래)일 때
        if d == 1 or d == 2:
            s %= (R - 1) * 2  # 상하 이동은 (R-1) * 2 주기를 가짐 (리밋 없이 계속 이동할 수 없도록 주기 처리)
            for _ in range(s):
                if nr == 0 and d == 1:  # 맨 위쪽에 도달하고 위로 이동할 경우, 아래로 방향 전환
                    d = 2
                elif nr == R-1 and d == 2:  # 맨 아래쪽에 도달하고 아래로 이동할 경우, 위로 방향 전환
                    d = 1
                nr += dr[d-1]  # 새 좌표로 이동

        # 좌우 이동(3: 오른쪽, 4: 왼쪽)일 때
        else:
            s %= (C - 1) * 2  # 좌우 이동은 (C-1) * 2 주기를 가짐
            for _ in range(s):
                if nc == 0 and d == 4:  # 맨 왼쪽에 도달하고 왼쪽으로 이동할 경우, 오른쪽으로 방향 전환
                    d = 3
                elif nc == C-1 and d == 3:  # 맨 오른쪽에 도달하고 오른쪽으로 이동할 경우, 왼쪽으로 방향 전환
                    d = 4
                nc += dc[d-1]  # 새 좌표로 이동

        # 상어가 이동한 후, 같은 위치에 상어가 있을 경우 크기가 큰 상어만 남도록 처리
        if (nr, nc) in new_sharks:
            if new_sharks[(nr, nc)][2] < z:  # 크기가 더 큰 상어가 있으면 기존 상어를 덮어씀
                new_sharks[(nr, nc)] = [s, d, z]
        else:
            new_sharks[(nr, nc)] = [s, d, z]  # 빈 칸이면 새로운 상어 정보를 저장

    return new_sharks  # 이동이 완료된 상어들의 새로운 상태 반환

# 낚시왕이 잡은 상어의 크기를 저장할 변수
total_size = 0

# 낚시왕이 1번 열부터 마지막 열까지 이동하며 상어를 잡는 과정
for king_pos in range(C):
    # 1번 열부터 차례대로 가장 가까운 상어를 잡는 과정
    for row in range(R):
        if (row, king_pos) in sharks:  # 현재 낚시왕이 있는 열에서 가장 가까운 상어를 찾음
            total_size += sharks[(row, king_pos)][2]  # 상어 크기를 합산
            del sharks[(row, king_pos)]  # 상어를 잡았으므로 해당 상어는 삭제
            break  # 한 마리만 잡을 수 있으므로 루프 종료

    # 상어들이 이동하는 과정
    sharks = move_sharks()

# 낚시왕이 잡은 상어 크기의 총합을 출력
print(total_size)
